File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.js
const fs = require("fs");
const stdin = (
	process.platform === "linux"
		? fs.readFileSync("/dev/stdin").toString()
		: fs.readFileSync("./input.txt").toString()
).split("\n");

const input = (() => {
	let line = 0;
	return () => stdin[line++];
})();

const [N, M] = input()
	.trim()
	.split(" ")
	.map((v) => parseInt(v));

const area = [];
for (let i = 0; i < N; i++) {
	area.push(
		input()
			.trim()
			.split("")
			.map((v) => parseInt(v))
	);
}

const direction = [
	[0, 1],
	[0, -1],
	[1, 0],
	[-1, 0],
];

function bfs() {
	const visited = Array.from(Array(N), () => Array.from(Array(M), () => Array(2).fill(0)));
	const queue = [[0, 0, 0]];
	visited[0][0][0] = 1;
	let count = 0;
	while (count < queue.length) {
		[x, y, c] = queue[count];

		if (x === N - 1 && y === M - 1) return visited[x][y][c];

		for (d of direction) {
			n_x = x + d[0];
			n_y = y + d[1];

			if (0 <= n_x && n_x < N && 0 <= n_y && n_y < M) {
				if (area[n_x][n_y] === 1 && c === 0) {
					visited[n_x][n_y][1] = visited[x][y][0] + 1;
					queue.push([n_x, n_y, 1]);
				}
				if (area[n_x][n_y] === 0 && visited[n_x][n_y][c] === 0) {
					visited[n_x][n_y][c] = visited[x][y][c] + 1;
					queue.push([n_x, n_y, c]);
				}
			}
		}
		count++;
	}
	return -1;
}

console.log(bfs());

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
import sys
from collections import deque


N, M = map(int, sys.stdin.readline().split())
area = [list(map(int, sys.stdin.readline().rstrip())) for _ in range(N)]
visited = [[[0] * 2 for m in range(M)] for n in range(N)]


def bfs():
    queue = deque([[0, 0, 0]])  # x, y, break
    visited[0][0][0] = 1
    direction = [[0, 1], [0, -1], [1, 0], [-1, 0]]

    while queue:
        [x, y, b] = queue.popleft()

        if x == N - 1 and y == M - 1:
            return visited[x][y][b]

        for d in direction:
            nx = x + d[0]
            ny = y + d[1]

            if 0 <= nx < N and 0 <= ny < M:
                if area[nx][ny] == 1 and b == 0:
                    visited[nx][ny][1] = visited[x][y][0] + 1
                    queue.append([nx, ny, 1])

                if area[nx][ny] == 0 and visited[nx][ny][b] == 0:
                    visited[nx][ny][b] = visited[x][y][b] + 1
                    queue.append([nx, ny, b])

    return -1


print(bfs())